### ISSUE TITLE
Replace GeoJSON with simpler immutable data structures

### ISSUE THREAD

**Issue by christophercliff** [Turfjs/turf#500]
I recently ran into a problem where I couldn't use turf because of the overhead of GeoJSON. I ended up porting turf functions to a form with simpler immutable data structures. This is a tenuous proposal, I realize it might be too niche for typical users.

**tl;dr GeoJSON is impractical in a complex application with large data set.**

## Problem

Suppose you have:

- large OpenStreetMap graph in memory
- React/Redux/Immutable.js app with immutable application state

For example https://christophercliff.com/dijkstra/.

Event handlers tend to look like:

1. Do calculation from immutable app state and OSM graph
2. Convert result to GeoJSON
3. Use GeoJSON in turf APIs
4. Convert result back to immutable app state
5. Render the result

Steps 2-4 are unnecessarily complex and can be prohibitively expensive.

## Proposal

Turf APIs should use low-level immutable objects. E.g. instead of `Feature` and `FeatureCollection`, turf APIs should accept and return `Immutable.List<Number>` and `Immutable.List<Immutable.List<Number>>`.

GeoJSON APIs could wrap the low-level APIs for back compatibility.

**Comment by tmcw** [comment]
> tl;dr GeoJSON is impractical in a complex application with large data set.

Having written geojson.io and Mapbox Studio, as well as many GeoJSON utilities, I think this is at the very least an overstatement.

Thanks for the suggestion, but my vote is 👎 .

1. "low-level immutable objects in JavaScript" don't exist. There are no language-level immutable objects in JavaScript, and libraries like Immutable.js, which I use and love, are low-level in no way - implementation, memory usage, conceptually, or whatever. They are high-level libraries with significant overhead. Worthwhile overhead, but no, not low-level. If this was written in a language with first-class immutability - and, trust me, I wish such a language was commonplace on the web, we'd use them. But it isn't, so we don't.
2. "Steps 2-4 are unnecessarily complex and can be prohibitively expensive." - are they? Sure, it's not a 1:1 conversion, and it would be nice for applications to do exactly the same as libraries, but then again, that isn't really possible given that libraries and applications are not a 1:1 relationship. But taking this more seriously: is conversion to and from Immutable.js _really_ the bottleneck in your application? Like, that one-time cost of moving back and forth, that isn't outweighed, in your case of using big datasets, by the efficiency of the algorithm, or by DOM manipulation, or something else?

**Comment by DenisCarriere** [comment]
For large complex applications I would recommend you use Vector Tiles. [Tippecanoe](https://github.com/mapbox/tippecanoe) is a well documented application to create Vector tiles from large GeoJSON features.

Using GeoJSON's with TurfJS is very well structured and I've personally managed large spatial datasets by only using TurfJS without adding any complexity to my source code.

I don't support this change :-1:
